File: iosApp/iosApp/Modules/NFTSend/NFTSendInteractor.swift
// Created by web3d4v on 04/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol NFTSendInteractor: AnyObject {
    var walletAddress: String? { get }
    func networkFees(network: Network) -> [NetworkFee]
    func fiatPrice(currency: Currency) -> Double
}

final class DefaultNFTSendInteractor {
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    
    init(
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
    }
}

extension DefaultNFTSendInteractor: NFTSendInteractor {
    
    var walletAddress: String? {
        try? networksService.wallet()?.address().toHexStringAddress().hexString
    }
    
    func networkFees(network: Network) -> [NetworkFee] {
        networksService.networkFees(network: network)
    }
    
    func fiatPrice(currency: Currency) -> Double {
        currencyStoreService.marketData(currency: currency)?.currentPrice?.doubleValue ?? 0
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/nftSend/NFTSendInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.nftSend

import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.types.NetworkFee
import com.sonsofcrypto.web3lib.types.toHexStringAddress

interface NFTSendInteractor {
    val walletAddress: String?
    fun networkFees(network: Network): List<NetworkFee>
    fun fiatPrice(currency: Currency): Double
}

class DefaultNFTSendInteractor(
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
): NFTSendInteractor {

    override val walletAddress: String? get() =
        networksService.wallet()?.address()?.toHexStringAddress()?.hexString

    override fun networkFees(network: Network): List<NetworkFee> =
        networksService.networkFees(network)

    override fun fiatPrice(currency: Currency): Double =
        currencyStoreService.marketData(currency)?.currentPrice ?: 0.toDouble()
}
